---
title: "Phylogenetic reconstruction"
author: Walter Xie and Alexei Drummond
output: md_document
---

## Preparation

In this lab, we will learn how to use R to construct phylogenetic trees using some basic algorithms,
such as UPGMA and Neighbour-joining, etc. 

Please download the nucleotide [alignment](ABCD.fasta), and save it into your working directory.
You can preview the alignment by using any text editor to open the file. 

We also need three R libraries: "ape", "phangorn", and "phytools". 
If your R has not got them, please use `install.packages` to install them.

## 1. Pairwise distances

Let's load "ape" first into R:

```{r}
library(ape)
```

Make sure the the alignment file is in your working directory, then load the alignment from the fasta file:

```{r}
# setwd("~/WorkSpace/BioSci700")
aln = read.FASTA("ABCD.fasta")
```

This alignment has 4 taxa and 20 sites. 
The nucleotides were composited to create the distance matrix in [Lecture 5](https://alexeidrummond.org/bayesian_phylo_lectures/lecture5/#/19).

Using the function `dist.dna` to compute the distribution of pairwise distances, 
we can multiple it with the number of sites (20) to obtain the matrix of pairwise distances.


```{r}
# 20 sites
d = dist.dna(aln, model = "raw") * 20
d
```

__Question 1 :__ what does this matrix tell us?

## 2. Algorithms

Secondly, we will use the library "phangorn" to create trees, and use "phytools". 

```{r}
library(phangorn)
library(phytools)
```

### 2.1 UPGMA

Now, please reuse the distance matrix `d` and compute the UPGMA tree. 
Use `plot` function to draw the tree, and use `edgelabels` to show branch lengths.

```{r}
treeUPGMA <- upgma(d)
plot(treeUPGMA, use.edge.length=T, no.margin=TRUE)
edgelabels(round(treeUPGMA$edge.length,2))
```


### 2.2 Neighbour-joining

Then compute the neighbour-joining tree. As it produces unrooted trees, 
you need to add the "unrooted" argurment to the `plot` function.

```{r}
treeNJ <- NJ(d)
plot(treeNJ, type="unrooted", use.edge.length=T, no.margin=TRUE)
edgelabels(treeNJ$edge.length)
```

__Question 2 :__ what is pros and cons of the neighbour-joining algorithm?


### 2.3 Parsimony

To understand the parsimony method, we need a more complex example.
Please follow the instructions below, which are modified from the "phangorn" [tutorial](https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html).

First load the alignment "primates", and remove the outgroup sequence "Bovine" to make the analysis simple.
The command `read.phyDat("???.fasta", format = "fasta", type = "DNA")` is used to load the file alternatively.

```{r}
fdir <- system.file("extdata/trees", package = "phangorn")
primates <- read.phyDat(file.path(fdir, "primates.dna"), format = "interleaved")
names(primates)
# rm "Bovine"
primates <- primates[-2]
names(primates)
```

We can now create the UPGMA and NJ tree respectively.

```{r}
dm  <- dist.ml(primates)
treeUPGMA  <- upgma(dm)
treeNJ  <- NJ(dm)
```

The UPGMA tree looks like:

```{r}
plot(treeUPGMA, no.margin=TRUE)
```

The NJ tree looks like:

```{r}
plot(treeNJ, type="unrooted", no.margin=TRUE)
```

Print the parsimony scores.

```{r}
parsimony(c(treeUPGMA, treeNJ), primates)
```

__Question 3 :__ Why are those two parsimony scores same? Assuming there are two trees giving different parsimony scores, which tree should be the maximum parsimony tree? 

We then can give a starting tree, and perform nearest-neighbor interchanges (NNI)
to find the maximum parsimony tree. 

```{r, results='hide'}
treePars <- optim.parsimony(treeUPGMA, primates, rearrangements = "NNI")
```

In the end, plot the maximum parsimony tree.

```{r}
plot(treePars, type="unrooted", no.margin=TRUE)
```


### 2.4 Maximum likelihood

The scripts below are using the maximum likelihood approach to 
estimate the parameters from "primates" data,
and eventually create a maximum likelihood phylogenetic tree from the inference.

First we create a NJ tree using F81 model as the starting tree.

```{r}
dm <- dist.ml(primates, "F81")
# NJ starting tree
tree <- NJ(dm)
```

Then we define the model to HKY + Γ(4) and optimise the parameters.

```{r, results='hide'}
# HKY + Γ(4)
fitStart <- pml(tree, primates, k=4)
fitHKY <- optim.pml(fitStart, model="HKY", optGamma=TRUE, rearrangement="stochastic")
```

The result is:

```{r}
fitHKY
```

Tips: you can use the command `str(fitHKY)` to list all available items inside "fitHKY". 

As you can see, the tree from this result is unrooted. 
So we will use the function `root` to assign the root to the tree given the outgroup taxon "Mouse". 

```{r}
is.rooted(fitHKY$tree) 
rt <- root(fitHKY$tree, outgroup = "Mouse", resolve.root = TRUE)
```

Finally, the maximum likelihood tree looks like:

```{r}
plot(rt, use.edge.length=T, no.margin=TRUE)
```

Add the command `edgelabels(round(rt$edge.length, 2))` to show all branch lengths.


### 2.5 Model selection

When you make a data analysis using the model-based approach, 
you need to know which model fits the best for the given data.

So we compare different nucleotide substitution models for "primates".

```{r, results='hide'}
mt = modelTest(primates)
```

```{r}
mt
```

As you can see, the "GTR+G+I" model has the best likelihood for "primates".
Therefore, we should choose "GTR+G+I" not "HKY+G" to run the phylogenetic analysis on this data.

## Assigment (15 marks)

Please download the [H5N1 data](H5N1.fasta), which is a subset of original dataset [Wallace et al., 2007](https://doi.org/10.1371/journal.pcbi.1000520).
It consists of 43 influenza A H5N1 hemagglutinin and neuraminidase gene sequences isolated from a variety of hosts 1996 - 2005 across sample locations.

Then make the following analyses: 

1. Use the R code that you have learnt to choose the "best" model for this data, and explain why you choose it. Please attach your R code with the report.

2. Install the latest version of [RAxML](https://github.com/stamatak/standard-RAxML/tags), and create the maximum likelihood tree by using the model that you choose in the 1st step. 
Please record your command to run the analysis, and visualize the tree in the report. 

   Tips: Use `-m` to define your model, and use `-#` or `-N` to specify multiple runs on distinct starting trees. It may take about several minutes for 10 runs on this data set using the single-threaded SSE3 version. You can plot the tree and colour the tips in R. But if you prefer to use a software with GUI, we recommend [FigTree](http://tree.bio.ed.ac.uk/software/figtree/). The mapping file [locations.txt](locations.txt) can be used to colour the tips. 

3. Write a reprot to discribe your analysis and explain the result, 
for example, what are the model and parameters, what the tree tells us,
and so on.


