---
title: "Phylogenetic reconstruction"
author: Walter Xie and Alexei Drummond
output: md_document
---

## Preparation

In this lab, we will learn how to use R to construct phylogenetic trees using some basic algorithms,
such as UPGMA and Neighbour-joining, etc. 

Please download the nucleotide [alignment](ABCD.fasta), and save it into your working directory.
You can preview the alignment by using any text editor to open the file. 

We also need three R libraries: "ape", "phangorn", and "phytools". 
If your R has not got them, please use `install.packages` to install them.

## 1. Pairwise distances

Let's load "ape" first into R:

```{r}
library(ape)
```

Make sure the the alignment file is in your working directory, then load the alignment from the fasta file:

```{r}
# setwd("~/WorkSpace/BioSci700")
aln = read.FASTA("ABCD.fasta")
```

This alignment has 4 taxa and 20 sites. 
The nucleotides were composited to create the distance matrix in [Lecture 5](https://alexeidrummond.org/bayesian_phylo_lectures/lecture5/#/19).

Using the function `dist.dna` to compute the distribution of pairwise distances, 
we can multiple it with the number of sites (20) to obtain the matrix of pairwise distances.


```{r}
# 20 sites
d = dist.dna(aln, model = "raw") * 20
d
```

__Question 1 :__ what does this matrix tell us?

## 2. Algorithms

Secondly, we will use the library "phangorn" to create trees, and use "phytools". 

```{r}
library(phangorn)
library(phytools)
```

### 2.1 UPGMA

Now, please reuse the distance matrix `d` and compute the UPGMA tree. 
Use `plot` function to draw the tree, and use `edgelabels` to show branch lengths.

```{r}
treeUPGMA <- upgma(d)
plot(treeUPGMA, use.edge.length=T, no.margin=TRUE)
edgelabels(round(treeUPGMA$edge.length,2))
```


### 2.2 Neighbour-joining

Then compute the neighbour-joining tree. As it produces unrooted trees, 
you need to add the "unrooted" argurment to the `plot` function.

```{r}
treeNJ <- NJ(d)
plot(treeNJ, type="unrooted", use.edge.length=T, no.margin=TRUE)
edgelabels(treeNJ$edge.length)
```

__Question 2 :__ what is pros and cons of the neighbour-joining algorithm?


### 2.3 Parsimony

To understand the parsimony method, we need a more complex example.
Please follow the instructions below, which are modified from the "phangorn" [tutorial](https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.html).

First load the alignment "primates", and remove the outgroup sequence "Bovine" to make the analysis simple.
The command `read.phyDat("???.fasta", format = "fasta", type = "DNA")` is used to load the file alternatively.

```{r}
primates <- read.phyDat("primates.fasta", format = "fasta")
names(primates)
```

We can now create the UPGMA and NJ tree respectively.

```{r}
dm  <- dist.ml(primates)
treeUPGMA  <- upgma(dm)
treeNJ  <- NJ(dm)
```

The UPGMA tree looks like:

```{r}
plot(treeUPGMA, no.margin=TRUE)
```

The NJ tree looks like:

```{r}
plot(treeNJ, type="unrooted", no.margin=TRUE)
```

Print the parsimony scores.

```{r}
parsimony(c(treeUPGMA, treeNJ), primates)
```

We then can give a starting tree, and perform nearest-neighbor interchanges (NNI)
to find the maximum parsimony tree. 

```{r, results='hide'}
treePars <- optim.parsimony(treeUPGMA, primates, rearrangements = "NNI")
```

In the end, plot the maximum parsimony tree.

```{r, fig.width = 6, fig.height = 6}
plot(treePars, type="unrooted", no.margin=TRUE)
```

### 2.4 Model selection

When you make a data analysis using the model-based approach, 
you need to know which model fits the best for the given data.

So we compare different nucleotide substitution models for "primates".

```{r, results='hide'}
mt = modelTest(primates)
```

```{r}
mt
```

As you can see, the "GTR+G" model has the best likelihood for "primates".
Therefore, we should choose it to run the phylogenetic analysis on this data.

### 2.5 Maximum likelihood

The scripts below are using the maximum likelihood approach to 
estimate the parameters from "primates" data,
and eventually create a maximum likelihood phylogenetic tree from the inference.

First we create a NJ tree using F81 model as the starting tree.

```{r}
dm <- dist.ml(primates, "F81")
# NJ starting tree
tree <- NJ(dm)
```

Then we define the model to GTR + Î“(4) as the model selection suggested and optimise the parameters.

```{r, results='hide'}
fitStart <- pml(tree, primates, k=4)
fitGTR <- optim.pml(fitStart, model="GTR", optGamma=TRUE, rearrangement="stochastic")
```

The result is:

```{r}
fitGTR
```

Tips: you can use the command `str(fitGTR)` to list all available items inside "fitGTR". 

As you can see, the result is an unrooted phylogenetic tree. 

```{r, fig.width = 6, fig.height = 6}
is.rooted(fitGTR$tree) 
plot(fitGTR$tree, type="unrooted")
```

If you want to root the tree, one simple method is to use the function `root` and specify the outgroup. 

```{r}
rt <- root(fitGTR$tree, outgroup = c("Tarsius_syrichta","Lemur_catta"), resolve.root = TRUE)
```

The rooted maximum likelihood tree looks like:

```{r}
plot(rt, use.edge.length=T, no.margin=TRUE)
```

Add the command `edgelabels(round(rt$edge.length, 2))` to show all branch lengths.



## Assigment (15 marks)

Please download the [H5N1 data](H5N1.fasta), which is a subset of original dataset [Wallace et al., 2007](https://doi.org/10.1073/pnas.0700435104).
It consists of 43 influenza A H5N1 hemagglutinin and neuraminidase gene sequences isolated from a variety of hosts 1996 - 2005 across sample locations.

Read the publication, and then make the following analyses: 

1. Use the R code that you have learnt to choose the "best" model for this data, and explain why you choose it. Please attach your R code with the report.

2. Install the latest version of [RAxML](https://github.com/stamatak/standard-RAxML/tags), and create the maximum likelihood tree by using the model that you choose in the 1st step.
You need to use the information provided in the publication to specify the outgroup taxon/taxa.
Please record your command to run the analysis, and visualize the tree in the report. 

   Tips: Use `-m` to define your model, and use `-#` or `-N` to specify multiple runs on distinct starting trees. It may take about several minutes for 10 runs on this data set using the single-threaded SSE3 version.
Use `-o` to specify the outgroup. More details are available in the manual.
         You can plot the tree and colour the tips in R. But if you prefer to use a software with GUI, we recommend [FigTree](http://tree.bio.ed.ac.uk/software/figtree/). The mapping file [locations.txt](locations.txt) can be used to colour the tips. 

3. Install another software [TempEst](http://tree.bio.ed.ac.uk/software/tempest/) to investigate the temporal signal at your ML tree.
Then find the "best-fitting root" and attach the visualized tree (coloured by locations) in the report.
Please record statistical results for both trees, the ML tree in step 2 and the best-fitting tree, 
and also explain the implication of these statistical results.

   Tips: Click "Parse Date" and use "prefix" `_` to load years from taxon labels into the program. 

4. Look at the "best-fitting" rooted tree, try to explain several important migration events according to the phylogeny. 

5. Write a full report to describe your analysis and explain the results, including the required answers from previous steps.


